xline/src/storage/kvwatcher.rs

```diff
@@ -392,37 +392,61 @@ where
     /// Create a new `Arc<KvWatcher>`
     pub(crate) fn new_arc(
         storage: Arc<KvStore<S>>,
-        mut kv_update_rx: mpsc::Receiver<(i64, Vec<Event>)>,
-        shutdown_trigger: Arc<event_listener::Event>,
+        kv_update_rx: mpsc::Receiver<(i64, Vec<Event>)>,
         sync_victims_interval: Duration,
+        shutdown_listener: shutdown::Listener,
     ) -> Arc<Self> {
         let watcher_map = Arc::new(RwLock::new(WatcherMap::new()));
         let kv_watcher = Arc::new(Self {
             storage,
             watcher_map,
         });
-        let victim_handle =
-            tokio::spawn(Arc::clone(&kv_watcher).sync_victims_task(sync_victims_interval));
-        let kv_updates_handle = tokio::spawn({
-            let kv_watcher = Arc::clone(&kv_watcher);
-            async move {
-                while let Some(updates) = kv_update_rx.recv().await {
+        let _victim_handle = tokio::spawn(Self::sync_victims_task(
+            Arc::clone(&kv_watcher),
+            sync_victims_interval,
+            shutdown_listener.clone(),
+        ));
+        let _kv_updates_handle = tokio::spawn(Self::kv_updates_task(
+            Arc::clone(&kv_watcher),
+            kv_update_rx,
+            shutdown_listener,
+        ));
+        kv_watcher
```

Comment:
Same as above. For example, we could collect these two handles and spawn a new task waiting for the shutdown event, and cancel the spawned tasks once the event is received. This could reduce the boilerplate code as long as the underlying task is cancel-safe.